scripts/apps/authoring-react/macros/macros.tsx

```diff
@@ -0,0 +1,213 @@
+import React from 'react';
+import {
+    IArticle,
+    IArticleSideWidget,
+    IExtensionActivationResult,
+    IFieldsV2,
+    IRestApiResponse,
+} from 'superdesk-api';
+import {gettext} from 'core/utils';
+import {AuthoringWidgetHeading} from 'apps/dashboard/widget-heading';
+import {AuthoringWidgetLayout} from 'apps/dashboard/widget-layout';
+import {httpRequestJsonLocal} from 'core/helpers/network';
+import {IMacro} from 'superdesk-interfaces/Macro';
+import {groupBy} from 'lodash';
+import {Button} from 'superdesk-ui-framework/react/components/Button';
+import {sdApi} from 'api';
+import {dispatchInternalEvent} from 'core/internal-events';
+import {generatePatch} from 'core/patch';
+import {ToggleBox} from 'superdesk-ui-framework/react/components/Togglebox';
+import {Switch} from 'superdesk-ui-framework/react/components/Switch';
+import {omitFields} from '../data-layer';
+import {Spacer} from 'core/ui/components/Spacer';
+import {nameof} from 'core/helpers/typescript-helpers';
+
+type IProps = React.ComponentProps<
+    IExtensionActivationResult['contributions']['authoringSideWidgets'][0]['component']
+>;
+
+const getLabel = () => gettext('Macros widget');
+
+interface IState {
+    macros: Array<IMacro> | null;
+    article: {
+        saved: IArticle | null,
+        autosaved: IArticle | null,
+    };
+    displayGrouped: boolean | null;
```

Comment:
When there is some special meaning like here, add a comment in the interface like
`displayGrouped: boolean | null; // null when grouping can't be toggled. It can't be toggled when there are no groups.`